### PR TITLE
Added shouldMatchResource, shouldNotMatchResource matchers (#2945)

### PR DESCRIPTION
**Pull request by bzareba**

### CODE REVIEW

File: kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/resource/matchers.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import io.kotest.matchers.shouldNot
+import java.io.File
+import java.net.URL
+import java.nio.file.Path
+import kotlin.contracts.contract
+import kotlin.io.path.createDirectories
+import kotlin.io.path.writeText
+
+infix fun String?.shouldMatchResource(path: String) {
+   contract {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/resource/matchers.kt]: fancy!

**Review by LeoColman — CHANGES_REQUESTED** [review]
What should happen when resources are binaries or any format other than text?
Can we add some test cases to cover this?

**Comment by bzareba** [comment]
@LeoColman Hi, I think we should let users compare string to any file. I've added test case to cover this. 

When it comes to binary files I think something like `ByteArray shouldMatchResource "/resource.bin"` would be a better fit.

**Comment by Kantis** [comment]
Can we let the user provide a matcher to compare actual with resource content with? For instance, what if the resource contains a json blob and you want to compare specific fields of the json.

Current implementation could also see issues due to line endings

**Comment by bzareba** [comment]
@Kantis I've added possibility to add custom matcher in form of matcher provider, so the usage is:
`givenValue.shouldMatchResource("/testResource.txt", ::startWith)`
or like this, enabling resource value modification before compare:
~`givenValue.shouldMatchResource("/testResource.txt") { s -> startWith(s.lowercase()) }`~

edit: I've added optional `ignoreLineSeparators` parameter so the syntax with lambda is:
`givenValue.shouldMatchResource("/testResource.txt", { s -> startWith(s.lowercase()) })`

**Comment by bzareba** [comment]
I also added possibility to ignore line separators differences. I made it default behavior.

**Comment by sksamuel** [comment]
@bzareba looks great. I think @LeoColman wanted a test on non-text resource and then this is good to merge.

**Comment by bzareba** [comment]
Hi @sksamuel I've added the test, but I think `ByteArray` matchers would be better with binary files, so I just added them :D I've pushed them here, please let me know if they should go in the separate PR

**Comment by bzareba** [comment]
I see there are some failing tests. That is probably due to autocrlf handling in git. I'll find a way to handle this, I'm thinking of creating the resources programmatically

**Comment by bzareba** [comment]
> I'm thinking of creating the resources programmatically

I forgot that `.gitattributes` can be used to keep original line-ending for these files. It should work now. :)

**Review by LeoColman — APPROVED** [review]

**Pull request merged**